infra/blueprint-test/pkg/krmt/krm.go

```diff
@@ -0,0 +1,312 @@
+	for _, s := range CommonSetters {
+		sKey, sVal, err := kpt.GenerateSetterKVFromEnvVar(s)
+		// if a common env var is not set, log and continue
+		if err != nil {
+			b.logger.Logf(b.t, "Skipping env var %s: %v", s, err)
+		} else {
+			b.logger.Logf(b.t, "Setting env var %s as setter %s: %s", s, sKey, sVal)
+			setters[sKey] = sVal
+		}
+	}
+	// merge user provided setters with env discovered setters
+	// user provided setters override env discovered setters
+	b.setters = kpt.MergeSetters(setters, b.setters)
+}
+
+// ValFromEnv returns value for a given env var, fails test if env var is not set.
+func (b *KRMBlueprintTest) ValFromEnv(k string) string {
+	v, err := utils.ValFromEnv(k)
+	if err != nil {
+		b.t.Fatal(err)
+	}
+	return v
+}
+
+// setupBuildDir prepares build dir with configs from bpDir.
+func (b *KRMBlueprintTest) setupBuildDir() {
+	// remove buildDir if exists
+	err := os.RemoveAll(b.buildDir)
+	if err != nil {
+		b.t.Fatalf("unable to remove %s :%v", b.buildDir, err)
+	}
+	// copy over configs into build dir
+	err = copy.Copy(b.bpDir, b.buildDir)
+	if err != nil {
+		b.t.Fatalf("unable to copy %s to %s :%v", b.bpDir, b.buildDir, err)
+	}
+	// subsequent kpt pkg update requires a clean git repo without uncommitted changes
+	// init a new git repo in build dir and commit changes
+	git := git.NewCmdConfig(b.t, git.WithDir(b.buildDir))
+	git.Init()
+	git.AddAll()
+	git.Commit()
+}
+
+// updateSetters updates existing setters with user provided setters.
+func (b *KRMBlueprintTest) updateSetters() {
+	rs, err := kpt.ReadPkgResources(b.buildDir)
+	if err != nil {
+		b.t.Fatalf("unable to read resources in %s :%v", b.buildDir, err)
+	}
+	kpt.UpsertSetters(rs, b.setters)
+	err = kpt.WritePkgResources(b.buildDir, rs)
+	if err != nil {
+		b.t.Fatalf("unable to write resources in %s :%v", b.buildDir, err)
+	}
+}
+
+// updatePkg updates a kpt pkg to a specified commit or the latest commit.
+func (b *KRMBlueprintTest) updatePkg() {
+	g := git.NewCmdConfig(b.t, git.WithDir(b.bpDir))
+	commit := b.updateCommit
+	if commit == "" {
+		commit = g.GetLatestCommit()
+	}
+	b.kpt.RunCmd("pkg", "update", fmt.Sprintf(".@%s", commit))
+
+}
+
+// DefaultInit sets up build directory, updates pkg, upserts setters and renders config.
+func (b *KRMBlueprintTest) DefaultInit(assert *assert.Assertions) {
+	b.setupBuildDir()
+	if b.updatePkgs {
+		b.updatePkg()
+	}
+	b.updateSetters()
+	kpt.NewCmdConfig(b.t, kpt.WithDir(b.buildDir)).RunCmd("fn", "render")
+}
+
+// DefaultApply installs resource-group, initializes inventory and apply pkg.
+func (b *KRMBlueprintTest) DefaultApply(assert *assert.Assertions) {
+	b.kpt.RunCmd("live", "install-resource-group")
+	b.kpt.RunCmd("live", "init")
+	b.kpt.RunCmd("live", "apply")
+}
+
+// DefaultVerify polls resource statuses until current.
+func (b *KRMBlueprintTest) DefaultVerify(assert *assert.Assertions) {
+	b.kpt.RunCmd("live", "status", "--output", "json", "--poll-until", "current", "--timeout", b.timeout)
```

Comment:
This should probably be moved into the apply behavior, since waiting for reconciliation is more akin to what happens during `terraform apply` (and I anticipate every test having to do this).

Reply:
Makes sense moved this to apply and added a helper to mimic default TF verify behavior (i.e assert that no resource changes after apply)